File: solutions/elixir/robot-simulator/4/lib/robot_simulator.ex
defmodule RobotSimulator do
  @directions %{:north => 0, :east => 1, :south => 2, :west => 3}
  defstruct position: {0, 0}, direction: :north

  @spec in_map(any) ::
          {:__block__ | {:., [], [:erlang | :not, ...]}, [],
           [{:= | {any, any, any}, [], [...]}, ...]}
  defguard in_map(direction)
           when direction in [:north, :south, :east, :west]

  defguard is_valid_position_tuple(position)
           when (is_tuple(position) and tuple_size(position) == 2)

  defguard is_valid_position_integer(a, b) when (is_integer(a) and is_integer(b))

  @doc """
  Create a Robot Simulator given an initial direction and position.

  Valid directions are: `:north`, `:east`, `:south`, `:west`
  """
  @spec create(direction :: atom, position :: {integer, integer}) :: %RobotSimulator{}
  def create(direction \\ :north, position \\ {0, 0})
  def create(direction, _position) when not in_map(direction), do: {:error, "invalid direction"}

  def create(_direction, position) when not is_valid_position_tuple(position),
    do: {:error, "invalid position"}

  def create(_direction, {a, b} = _position) when not is_valid_position_integer(a, b),
    do: {:error, "invalid position"}

  def create(direction, position) do
    %RobotSimulator{direction: direction, position: position}
  end

  @doc """
  Simulate the robot's movement given a string of instructions.

  Valid instructions are: "R" (turn right), "L", (turn left), and "A" (advance)
  """
  @spec simulate(robot :: any, instructions :: String.t()) :: any
  def simulate(robot, instructions) do
    simulate_helper(robot, to_charlist(instructions))
  end

  defp simulate_helper(robot, []), do: robot

  defp simulate_helper(robot, [?R | tail] = _instructions),
    do:
      simulate_helper(
        %{
          robot
          | direction: enum_map_helper(robot.direction, ?R)
        },
        tail
      )

  defp simulate_helper(robot, [?L | tail] = _instructions),
    do:
      simulate_helper(
        %{
          robot
          | direction: enum_map_helper(robot.direction, ?L)
        },
        tail
      )

  defp simulate_helper(robot, [?A | tail] = _instructions),
    do: simulate_helper(advance_helper(robot), tail)

  defp simulate_helper(_robot, _instructions), do: {:error, "invalid instruction"}

  defp enum_map_helper(direction, ?R) do
    key_find(rem(@directions[direction] + 1, 4))
  end

  defp enum_map_helper(direction, ?L) do
    key_find(rem(@directions[direction] + -1 + 4, 4))
  end

  defp key_find(val) do
    {key, _} = Enum.find(@directions, fn {_, value} -> val == value end)
    key
  end

  defp advance_helper(%RobotSimulator{direction: :north, position: {posx, posy}} = robot),
    do: %{robot | position: {posx, posy + 1}}

  defp advance_helper(%RobotSimulator{direction: :east, position: {posx, posy}} = robot),
    do: %{robot | position: {posx + 1, posy}}

  defp advance_helper(%RobotSimulator{direction: :south, position: {posx, posy}} = robot),
    do: %{robot | position: {posx, posy - 1}}

  defp advance_helper(%RobotSimulator{direction: :west, position: {posx, posy}} = robot),
    do: %{robot | position: {posx - 1, posy}}

  @doc """
  Return the robot's direction.

  Valid directions are: `:north`, `:east`, `:south`, `:west`
  """
  @spec direction(robot :: %RobotSimulator{}) :: atom
  def direction(robot) do
    robot.direction
  end

  @doc """
  Return the robot's position.
  """
  @spec position(robot :: %RobotSimulator{}) :: {integer, integer}
  def position(robot) do
    robot.position
  end
end

File: solutions/erlang/robot-simulator/4/src/robot_simulator.erl
-module(robot_simulator).

-behaviour(gen_server).

-export([init/1, handle_call/3, handle_cast/2]).
-export([advance/1, create/0, direction/1, left/1, place/3, position/1, right/1]).

advance(Robot) ->
    gen_server:call(Robot, advance).

create() ->
    {ok, Id} = gen_server:start(?MODULE, [#{position => {0, 0}, direction => north}], []),
    Id.

direction(Robot) ->
    gen_server:call(Robot, direction).

left(Robot) ->
    gen_server:call(Robot, left).

place(Robot, Direction, Position) ->
    gen_server:call(Robot, {place, #{position => Position, direction => Direction}}).

position(Robot) ->
    gen_server:call(Robot, position).

right(Robot) ->
    gen_server:call(Robot, right).

%server methods
init([RobotState | _Tail]) ->
    {ok, RobotState}.

handle_call(left, _From, RobotState) ->
    case maps:get(direction, RobotState) of
        north ->
            {reply, ok, maps:update(direction, west, RobotState)};
        east ->
            {reply, ok, maps:update(direction, north, RobotState)};
        south ->
            {reply, ok, maps:update(direction, east, RobotState)};
        west ->
            {reply, ok, maps:update(direction, south, RobotState)}
    end;
handle_call(right, _From, RobotState) ->
    case maps:get(direction, RobotState) of
        north ->
            {reply, ok, maps:update(direction, east, RobotState)};
        east ->
            {reply, ok, maps:update(direction, south, RobotState)};
        south ->
            {reply, ok, maps:update(direction, west, RobotState)};
        west ->
            {reply, ok, maps:update(direction, north, RobotState)}
    end;
handle_call(advance, _From, RobotState) ->
    {X, Y} = maps:get(position, RobotState),
    case maps:get(direction, RobotState) of
        north ->
            {reply, ok, maps:update(position, {X, Y + 1}, RobotState)};
        east ->
            {reply, ok, maps:update(position, {X + 1, Y}, RobotState)};
        south ->
            {reply, ok, maps:update(position, {X, Y - 1}, RobotState)};
        west ->
            {reply, ok, maps:update(position, {X - 1, Y}, RobotState)}
    end;
handle_call({place, RobotState_new}, _From, _RobotState) ->
    {reply, ok, RobotState_new};
handle_call(direction, _From, RobotState) ->
    {reply, maps:get(direction, RobotState), RobotState};
handle_call(position, _From, RobotState) ->
    {reply, maps:get(position, RobotState), RobotState}.

%handle other messages.
handle_cast(_Msg, N) ->
    {noreply, N}.
